repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go

```diff
@@ -0,0 +1,288 @@
+)
+
+type IpfsFetcher struct {
+	distPath string
+	limit    int64
+	peers    []peer.AddrInfo
+
+	openOnce  sync.Once
+	openErr   error
+	closeOnce sync.Once
+	closeErr  error
+
+	ipfs         iface.CoreAPI
+	ipfsTmpDir   string
+	ipfsStopFunc func()
+}
+
+// NewIpfsFetcher creates a new IpfsFetcher
+//
+// Specifying "" for distPath sets the default IPNS path.
+// Specifying 0 for fetchLimit sets the default, -1 means no limit.
+func NewIpfsFetcher(distPath string, fetchLimit int64, peers []peer.AddrInfo) *IpfsFetcher {
+	f := &IpfsFetcher{
+		limit:    defaultFetchLimit,
+		distPath: migrations.LatestIpfsDist,
+		peers:    peers,
+	}
+
+	if distPath != "" {
+		if !strings.HasPrefix(distPath, "/") {
+			distPath = "/" + distPath
+		}
+		f.distPath = distPath
+	}
+
+	if fetchLimit != 0 {
+		if fetchLimit < 0 {
+			fetchLimit = 0
+		}
+		f.limit = fetchLimit
+	}
+
+	return f
+}
+
+// Fetch attempts to fetch the file at the given path, from the distribution
+// site configured for this HttpFetcher.  Returns io.ReadCloser on success,
+// which caller must close.
+func (f *IpfsFetcher) Fetch(ctx context.Context, filePath string) (io.ReadCloser, error) {
+	// Initialize and start IPFS node on first call to Fetch, since the fetcher
+	// may be created by not used.
+	f.openOnce.Do(func() {
+		f.ipfsTmpDir, f.openErr = initTempNode(ctx)
+		if f.openErr != nil {
+			return
+		}
+
+		f.ipfs, f.ipfsStopFunc, f.openErr = startTempNode(f.ipfsTmpDir, f.peers)
+	})
+
+	fmt.Printf("Fetching with IPFS: %q\n", filePath)
+
+	if f.openErr != nil {
+		return nil, f.openErr
+	}
+
+	iPath, err := parsePath(path.Join(f.distPath, filePath))
+	if err != nil {
+		return nil, err
+	}
+
+	nd, err := f.ipfs.Unixfs().Get(ctx, iPath)
+	if err != nil {
+		return nil, err
+	}
+
+	fileNode, ok := nd.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("%q is not a file", filePath)
+	}
+
+	if f.limit != 0 {
+		return migrations.NewLimitReadCloser(fileNode, f.limit), nil
+	}
+	return fileNode, nil
+}
+
+func (f *IpfsFetcher) Close() error {
+	f.closeOnce.Do(func() {
+		if f.ipfsStopFunc != nil {
+			// Tell ipfs node to stop and wait for it to stop
+			f.ipfsStopFunc()
+		}
+
+		if f.ipfsTmpDir != "" {
+			// Remove the temp ipfs dir
+			f.closeErr = os.RemoveAll(f.ipfsTmpDir)
+		}
+	})
+	return f.closeErr
+}
+
+func initTempNode(ctx context.Context) (string, error) {
+	err := setupPlugins()
+	if err != nil {
+		return "", err
+	}
+
+	identity, err := config.CreateIdentity(ioutil.Discard, []options.KeyGenerateOption{
+		options.Key.Type(options.Ed25519Key),
+	})
+	if err != nil {
+		return "", err
+	}
+	cfg, err := config.InitWithIdentity(identity)
+	if err != nil {
+		return "", err
+	}
+
+	// create temporary ipfs directory
+	dir, err := ioutil.TempDir("", "ipfs-temp")
+	if err != nil {
+		return "", fmt.Errorf("failed to get temp dir: %s", err)
+	}
+
+	// configure the temporary node
+	cfg.Routing.Type = "dhtclient"
+
+	// Disable listening for inbound connections
```

Comment:
We could do that, but then if you try and import the data into the new node from the old node for seeding purposes you're going to need to either:
- Use the import/export APIs (e.g. use CAR files to move the data around)
- Have the temporary node connect to the new node so it can pull the data using Bitswap

Reply:
Since the files are already available on local disk (since they need to be run), it is just as easy for me to add the files to the real node after migration. So, I think this is fine since I am not using the temp node for anything other than downloading migrations.

Reply:
> it is just as easy for me to add the files to the real node after migration

Right, but if we do this it needs to be as a CAR file not a UnixFS import of some binary since we want the CIDs to match and `ipfs add <file>` is not guaranteed to be the same across versions and boxing ourselves into some parameters now wouldn't be a good idea either. i.e. go-ipfs needs to understand "I am importing this DAG" not "I am importing this binary", whether it's a CAR file via DAG import, or UnixFS via Bitswap is 🤷.

Reply:
- When migrations are downloaded via IPFS, add the migration archives to IPFS by connecting to the temporary migration node and getting the migrations over IPFS.
- When migrations are downloaded via HTTP, add the migration archive files directly to IPFS.  After talking to @Stebalien he thought it was still worth to add the files directly, even though it might result in a different dag.  If we decide later to not do that, it is a trivial change.